Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
+                    add(licenseJarPath);
+                }});
+            }
+
+            URL[] urls = new URL[] {commonJarFile.toURI().toURL(), jxBrowserJarFile.toURI().toURL(), licenseJarFile.toURI().toURL()};
+
+            ClassLoader classLoader = new URLClassLoader(urls, JxBrowserUtil.class.getClassLoader());
+
+            Class<?> browserClass = Class.forName(JXBROWSER_CLASS_BROWSER, true, classLoader);
             Object browser = browserClass.newInstance();
-            Class<?> browserViewClass = Class.forName("com.teamdev.jxbrowser.chromium.swing.BrowserView");
+            Class<?> browserViewClass = Class.forName(JXBROWSER_CLASS_BROWSERVIEW, true, classLoader);
             Constructor<?> browserViewConstructor = browserViewClass.getConstructor(browserClass);
             browserView = browserViewConstructor.newInstance(browser);
 
             if (browserView != null) {
                 Method loadURLMethod = browserClass.getMethod("loadURL", String.class);
-                loadURLMethod.invoke(url);
+                loadURLMethod.invoke(browser, url);
             }
         } catch (Exception e) {
-            throw new Exception("Fail to load JxBrowser or load URL: " + e.getMessage());
+            throw new JxBrowserException("Fail to load JxBrowser or load URL: " + e.getMessage());
         }
 
         return (JComponent)browserView;
     }
+
+    /**
+     * Download JxBrowser jars (i.e. jxbrowser-win32-6.16.jar and license.jar) and check digest (with retries) into given targetPath
+     * @param filePathNames
+     * @return
+     */
+    private static boolean downloadFilesFromAzure(@NotNull List<Path> filePathNames) {
+        boolean result = true;
+        ExecutorService threadPool = Executors.newFixedThreadPool(2);
+        try {
+            downloadAndCheckDigest(threadPool, filePathNames);
+
+            threadPool.shutdown();
+            threadPool.awaitTermination((long) (MAX_WAIT_TIME_IN_SECONDS * Math.pow(RETRY_BACKOFF_FACTOR, MAX_RETRY_TIMES)), TimeUnit.SECONDS);
+        } catch (Exception ex) {
+            log.severe("Download JxBrowser jars failed with exception " + ex.getMessage());
+            result = false;
+        } finally {
+            threadPool.shutdownNow();
+        }
+
+        return result;
+    }
+
+    /**
+     * Check the file correctness by comparing the MD5 digest with the digest file. Return true if both file and digest exist and match.
+     * @param filePathName
+     * @param digestFilePathName
+     * @return
+     */
+    private static boolean checkFileDigest(@NotNull Path filePathName, @NotNull Path digestFilePathName) {
+        File targetFile = filePathName.toFile();
+        if (!targetFile.exists() || !targetFile.isFile()) {
+            log.warning(filePathName + " doesn't exist");
+            return false;
+        }
+
+        File digestFile = digestFilePathName.toFile();
+        if (!digestFile.exists() || !digestFile.isFile()) {
```

Comment:
>digestFile.isFile() [](start = 37, length = 19)

isFile() will return false if a regular file not exists.

Reply:
Yes. However prefer to keep it as it is since this should not be performance bottleneck and it makes the logic look clear

---
In reply to: [156554195](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156554195) [](ancestors = 156554195)